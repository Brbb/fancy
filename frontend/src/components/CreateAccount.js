import React, { Component } from "react";
import "./CreateAccount.css";
import auth from "../services/auth/api";

class CreateAccount extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      password: "",
      repeatPassword: "",
      error: ""
    };
  }

  dismissError = () => {
    this.setState({ error: "" });
  };

  handleUserChange = event => {
    this.setState({ username: event.target.value });
  };
  handlePassChange = event => {
    this.setState({ password: event.target.value });
  };
  handleRepeatPassChange = event => {
    if (this.state.password !== event.target.value) {
      this.setState({ error: "Passwords must match" });
    } else this.dismissError();
    this.setState({ repeatPassword: event.target.value });
  };

  signup = async () => {
    var signupResult = await auth.signup(
      this.state.username,
      this.state.password,
      this.state.repeatPassword
    );
    if (signupResult.success) {
      this.props.history.push("/");
    } else this.setState({ error: signupResult.reason });
  };

  render() {
    return (
      <div className="CreateAccount formContainer">
        <form className="form" onSubmit={this.handleSubmit}>
          <label className="formTitle">Fancy App</label>
          <label>Email</label>
          <input
            type="text"
            className="textField"
            placeholder="user@email.com"
            data-test="username"
            value={this.state.username}
            onChange={this.handleUserChange}
          />

          <label>Password</label>
          <input
            type="password"
            placeholder="super secret"
            data-test="password"
            className="textField"
            value={this.state.password}
            onChange={this.handlePassChange}
          />
          <input
            type="password"
            placeholder="repeat super secret"
            data-test="password"
            className="textField"
            value={this.state.repeatPassword}
            onChange={this.handleRepeatPassChange}
          />
          {this.state.error && (
            <label
              className="loginError"
              data-test="error"
              onClick={this.dismissError}
            >
              {this.state.error}
            </label>
          )}
          <input
            className="submitBtn"
            type="button"
            value="Create New Account"
            onClick={this.signup}
            disabled={
              this.state.username === "" ||
              this.state.password === "" ||
              this.state.repeatPassword === "" ||
              this.state.error !== ""
            }
          />
        </form>
      </div>
    );
  }
}

export default CreateAccount;
