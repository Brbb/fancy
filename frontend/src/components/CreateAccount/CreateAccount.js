import React, { Component } from "react";
import "./CreateAccount.css";
import auth from "../../services/auth/api";
import { Button, InputField, Message, SectionTitle } from "../Elements/Basics";
import PasswordGroup from "../Elements/PasswordGroup";

class CreateAccount extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      password: "",
      repeatPassword: "",
      error: ""
    };
  }

  handleUserChange = event => {
    this.setState({ username: event.target.value });
  };

  signup = async () => {
    var result = await auth.signup(
      this.state.username,
      this.state.password,
      this.state.repeatPassword
    );
    if (!result.err) {
      this.props.history.push("/");
    } else this.setState({ error: result.err });
  };

  render() {
    return (
      <div className="f-section bordered centered">
        <SectionTitle text="New Account" />
        <InputField
          name="emailField"
          placeholder="user@email.com"
          value={this.state.username}
          onChange={this.handleUserChange}
          text="Email"
        />
        <PasswordGroup
          onPasswordsChange={(newPassword, repeatPassword, error) => {
            this.setState({
              password: newPassword,
              repeatPassword: repeatPassword,
              error: error
            });
          }}
        />
        <Button
          name="create-button"
          className="f-submit-button"
          text="Create New Account"
          onClick={this.signup}
          disabled={
            this.state.username === "" ||
            this.state.password === "" ||
            this.state.repeatPassword === "" ||
            this.state.error !== ""
          }
        />
        <Button
          className="f-button"
          text="← Back"
          onClick={() => this.props.history.push("/")}
        />
        <Message text={this.state.error} success={this.state.error === ""} />
      </div>
    );
  }
}

export default CreateAccount;
