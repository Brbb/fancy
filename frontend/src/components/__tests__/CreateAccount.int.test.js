import CreateAccount from "../CreateAccount/CreateAccount";
import axios from "axios";
jest.mock("axios");

const mockUser = email => {
  return {
    data: {
      id: "asd123",
      email: email,
      settings: {
        language: "English",
        isPrivate: false
      }
    }
  };
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe("CreateAccount component, integration test", () => {
  describe("Account creation", () => {
    test("User types email and passwords, clicks on create and gets a user as result.", () => {
      let location = { pathname: "/new" };
      let historyMock = {
        push: jest.fn(path => {
          location.pathname = path;
        })
      };

      jest.spyOn(axios, "post").mockImplementation((url, body) => {
        if (url === "/api/auth/signup")
          return Promise.resolve(mockUser(body.email));
      });

      // User fills the form
      let wrapper = mount(<CreateAccount history={historyMock} />);
      wrapper
        .find('[name="emailField"]')
        .last()
        .simulate("change", { target: { value: "test@email.com" } });
      wrapper
        .find('[name="newPasswordField"]')
        .last()
        .simulate("change", { target: { value: "pass123" } });
      wrapper
        .find('[name="repeatNewPasswordField"]')
        .last()
        .simulate("change", { target: { value: "pass123" } });

      let createButton = wrapper.find('[name="create-button"]').first();

      expect(wrapper).toMatchSnapshot(); // fields are populated and button is ready to be clicked
      expect(wrapper.state().username).toEqual("test@email.com");
      expect(wrapper.state().password).toEqual("pass123");
      expect(wrapper.state().repeatPassword).toEqual("pass123");
      expect(createButton.prop("disabled")).toBeFalsy();
      expect(axios.post).toBeCalledTimes(0);

      // submit
      return createButton
        .prop("onClick")()
        .then(() => {
          // we called the signup API
          expect(axios.post).toBeCalledTimes(1);
          // successfully redirected to Login
          expect(location.pathname).toBe("/");
        });
    });

    test("User types existing email, clicks on create and gets error message.", () => {
      jest.spyOn(axios, "post").mockImplementation(url => {
        if (url === "/api/auth/signup") {
          return Promise.reject({
            response: { data: { err: "Email exists" } }
          });
        }
      });

      // User fills the form
      let wrapper = mount(<CreateAccount />);
      let createButton = wrapper.find('[name="create-button"]').first();

      // submit
      return createButton
        .prop("onClick")()
        .then(err => {
          // we called the signup API
          expect(axios.post).toBeCalledTimes(1);
          // sets the error message to err
          expect(wrapper.state().error).toBe("Email exists");
        });
    });
  });
});
