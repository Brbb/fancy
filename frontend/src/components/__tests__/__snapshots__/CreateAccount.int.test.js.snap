// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount component, integration test Account creation User types email and passwords, clicks on create and gets a user as result. 1`] = `
<div
  className="f-section bordered centered"
>
  <div
    className="f-section-title"
  >
    <label>
      New Account
    </label>
  </div>
  <div
    className="f-input"
  >
    <label
      className="f-input-field-label"
    >
      Email
    </label>
    <input
      className="f-input-field"
      name="emailField"
      onChange={[Function]}
      placeholder="user@email.com"
      value="test@email.com"
    />
  </div>
  <div>
    <div
      className="f-input"
    >
      <label
        className="f-input-field-label"
      >
        Password
      </label>
      <input
        className="f-input-field"
        name="newPasswordField"
        onChange={[Function]}
        placeholder="Super Secret"
        type="password"
        value="pass123"
      />
    </div>
    <div
      className="f-input"
    >
      <label
        className="f-input-field-label"
      />
      <input
        className="f-input-field"
        name="repeatNewPasswordField"
        onChange={[Function]}
        placeholder="Repeat Super Secret"
        type="password"
        value="pass123"
      />
    </div>
  </div>
  <input
    className="f-submit-button"
    disabled={false}
    name="create-button"
    onClick={[Function]}
    type="button"
    value="Create New Account"
  />
  <input
    className="f-button"
    onClick={[Function]}
    type="button"
    value="← Back"
  />
  <label
    className="f-message f-message-s"
  />
</div>
`;
